### DISCUSSION TITLE
Dropdown of activity property in Custom Activity

### DISCUSSION THREAD

**Discussion by makattar** [elsa-workflows/elsa-core#1246] [Q&A] [answered]
Hi ,
I am creating a custom activity where i wants to create a activity property.That activity property will have a value of type string.But I wants to show as dropdown in designer where user selects values from dropdown.

**Comment by sfmskywalker** [comment, marked as answer]
You can achieve that by applying the `"dropdown"` UI hint. This is what the **Send HTTP Request** does:

```csharp
/// <summary>
/// The HTTP method to use.
/// </summary>
[ActivityInput(
  UIHint = ActivityInputUIHints.Dropdown,
  Hint = "The HTTP method to use when making the request.",
  Options = new[] { "GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "HEAD" },
  SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid }
)]
public string? Method { get; set; }
```

Notice that the `Method` property is of type `string` and that a finite set of options are provided from which the user can select one.
If you have a dynamic set of options that you want to provide at runtime, use the `OptionsProvider` property instead, which must be a `Type` that implements `IActivityPropertyOptionsProvider`, from which you return a collection of either `string` or `SelectListItem` objects.

For example:

```csharp
[ActivityInput(
  UIHint = ActivityInputUIHints.Dropdown,
  Hint = "The HTTP method to use when making the request.",
  OptionsProvider = typeof(MyOptionsProvider),
  SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid }
)]
public string? Method { get; set; }

public class MyOptionsProvider : IActivityPropertyOptionsProvider
{
   public object GetOptions(PropertyInfo property) => new[] { "GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "HEAD" };
}
```

**Reply by srihariraovadlani** [reply]
GetOptions method is being called when the process is initialized, i would need to refresh the data by calling API whenever designer is opened.  understood that IRuntimeSelectListItemsProvider is solution for it. Can you please share code snippet on how to use it. 

Appreciate your help!

**Reply by makattar** [reply]
I don't know exactly answer of it because I tried hard-coded string there But I think you can use asyn await syntax of c#.Just put that async code (Call to api) assigned to getoptions method.

**Comment by makattar** [comment]
Can you please tell me in which namespace **ActivityInput** Resides.Intellisense not giving any suggestions there.

**Reply by sfmskywalker** [reply]
Ah sorry, `ActivityInputPropertyAttribute` exists only in 2.1. In 2.0, the attribute is called `ActivityPropertyAttribute`.

I don’t know off hand the namespace (not sitting behind a big screen yet), but you should be able to find it now. If not, let me know.

**Reply by makattar** [reply]
Ok sir...I will try it...Thank you

**Reply by makattar** [reply]
I tried to implement it but It shows error messages on  UIHint = ActivityInputUIHints.Dropdown . Means ActivityInputUIHints not defined.

**Reply by sfmskywalker** [reply]
Can you try [ActivityPropertyUIHints ](https://github.com/elsa-workflows/elsa-core/blob/ae848a90f8865b09299e989674a6afc68da3d301/src/core/Elsa.Abstractions/Design/ActivityPropertyUIHints.cs) instead?

**Reply by makattar** [reply]
Thanks Sir.It worked.
Please also help me on https://github.com/elsa-workflows/elsa-core/discussions/1248
